File: Coursework/Numerical Analysis/cholesky.py
import numpy as np

def cholesky(A: np.ndarray) -> np.ndarray:
    '''    
    Cholesky decomposition of a symmetric positive definite matrix A

    parameters:
        A: symmetric positive definite matrix
        
    returns: 
        lower triangular matrix L
    '''


    n = len(A)
    L = np.zeros((n,n))
    for i in range(n):
        for j in range(i+1):
            tmp_sum = 0
            for k in range(j):
                tmp_sum += L[i,k] * L[j,k]
            if (i == j):
                L[i,j] = np.sqrt(A[i,i] - tmp_sum)
            else:
                L[i,j] = (A[i,j] - tmp_sum) / L[j,j]
    return L

File: Coursework/Numerical Analysis/Methods_First_Exam/cholesky.jl
function cholesky(A)
    n = size(A,1)
    L = zeros(n,n)
    for j = 1:n
        for i = j:n
            if i == j
                L[i,j] = sqrt(A[i,j] - sum(L[i,1:j-1].^2))
            else
                L[i,j] = (A[i,j] - sum(L[i,1:j-1].*L[j,1:j-1]))/L[j,j]
            end
        end
    end
    return L
end